### PR TITLE
Make HostMemoryBuffer call into the DefaultHostMemoryAllocator

### PR DESCRIPTION
**Pull request by revans2**
## Description
This is step 3 in a process of getting java host memory allocation to be plugable under a single allocation API. This is really only used for large memory allocations, which is what matters.

This changes the most common java host memory allocation API to call into the plugable host memory allocation API. The reason that this had to be done in multiple steps is because the Spark Plugin code was calling into the common memory allocation API, and memory allocation would end up calling itself recursively.

Step 1. Create a new API that will not be called recursively (https://github.com/rapidsai/cudf/pull/17197)
Step 2. Have the Java plugin use that new API instead of the old one to avoid any recursive invocations (https://github.com/NVIDIA/spark-rapids/pull/11671)
Step 3. Update the common API to use the new backend (this)

There are likely to be more steps after this that involve cleaning up and removing APIs that are no longer needed.

This is marked as breaking even though it does not break any APIs, it changes the semantics enough that it feels like a breaking change.

This is blocked and should not be merged in until Step 2 is merged in, to avoid breaking the Spark plugin.

## Checklist
- [X] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [X] New or existing tests cover these changes.
- [X] The documentation is up to date with these changes.

### CODE REVIEW

**Review by ttnghia — APPROVED** [review]

**Review by abellina — APPROVED** [review]

**Comment by revans2** [comment]
Even though this is approved and https://github.com/NVIDIA/spark-rapids/pull/11671 was just merged in I am going to hold off on merging this until https://github.com/NVIDIA/spark-rapids-jni/pull/2561 is also merged in. On the surface it should not be related, but the error only presented itself after this was merged in. I think it changed the timing of a race condition to cause the problem, but I am not 100% sure on that.

**Comment by revans2** [comment]
All of the dependencies are in and we should be good to go with this final change.

**Pull request merged**
